app/test/vesting_test.go

```diff
@@ -0,0 +1,674 @@
+	vestingAmount        = testfactory.BaseAccountDefaultBalance
+)
+
+type accountDispenser struct {
+	names   []string
+	counter int
+}
+
+type accountType int
+
+const (
+	RegularAccountType accountType = iota + 1
+	DelayedVestingAccountType
+	PeriodicVestingAccountType
+	ContinuousVestingAccountType
+)
+
+type VestingModuleTestSuite struct {
+	suite.Suite
+
+	accounts    sync.Map // map[_accountType]_accountDispenser
+	accountsMut sync.Mutex
+
+	kr   keyring.Keyring
+	cctx testnode.Context
+	ecfg encoding.Config
+}
+
+func TestVestingModule(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping Vesting accounts test in short mode.")
+	}
+	suite.Run(t, new(VestingModuleTestSuite))
+}
+
+func (s *VestingModuleTestSuite) SetupSuite() {
+	t := s.T()
+	t.Log("setting up vesting module test suite")
+
+	s.kr = testfactory.GenerateKeyring()
+
+	genOpts := []testnode.GenesisOption{}
+	genOpts = append(genOpts, s.initRegularAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initDelayedVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initPeriodicVestingAccounts(totalAccountsPerType))
+	genOpts = append(genOpts, s.initContinuousVestingAccounts(totalAccountsPerType))
+
+	s.startNewNetworkWithGenesisOpt(genOpts...)
+}
+
+// startNewNetworkWithGenesisOpt creates a new test network with the specified genesis options for the VestingModuleTestSuite.
+// It initializes a default Tendermint configuration (tmCfg) and default application configuration (appConf).
+// The target block time is set to 1 millisecond. It applies the given genesis options to the test network
+// and stores the created client context (cctx) in the VestingModuleTestSuite.
+// The keyring of the context is set to the keyring (s.kr) of the VestingModuleTestSuite.
+func (s *VestingModuleTestSuite) startNewNetworkWithGenesisOpt(genesisOpts ...testnode.GenesisOption) {
+	s.ecfg = encoding.MakeConfig(app.ModuleEncodingRegisters...)
+	cfg := testnode.DefaultConfig().WithGenesisOptions(testnode.ImmediateProposals(s.ecfg.Codec))
+	cfg.GenesisOptions = genesisOpts
+
+	cctx, _, _ := testnode.NewNetwork(s.T(), cfg)
+	s.cctx = cctx
+	s.cctx.Keyring = s.kr
+}
+
+func (s *VestingModuleTestSuite) TestGenesisDelayedVestingAccountsSpendableBalanceUnlocked() {
+	block, err := s.cctx.LatestBlock()
+	assert.NoError(s.T(), err)
+	// We need to wait for a block because sometimes querying based on the
+	// latest height throws an SDK error saying the height is in the future
+	assert.NoError(s.T(), s.cctx.WaitForNextBlock())
+
+	// find an unlocked delayed vesting account.
+	vAcc, name, err := s.getAnUnusedDelayedVestingAccount(0)
+	assert.NoError(s.T(), err)
+
+	address := getAddress(name, s.cctx.Keyring).String()
+	balances, err := testfactory.GetAccountSpendableBalanceByBlock(s.cctx.GRPCClient, address, block)
+	assert.NoError(s.T(), err)
+
+	expectedSpendableBal := vAcc.GetVestedCoins(block.Time).AmountOf(app.BondDenom).Int64() + initBalanceForGasFee
+	assert.EqualValues(s.T(),
```

Comment:
[question] should this additionally verify that expectedSpendableBal > initBalanceForGasFee to avoid the case mentioned above:

> [question] It seem possible for getAnUnusedDelayedVestingAccount to return a delayed vesting account whose start time and end time is in the future. If that's the case, this test doesn't verify that any tokens are unlocked 🤔

Reply:
this is done in the new structure